core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenRevocationAuthenticationProvider.java

```diff
@@ -0,0 +1,107 @@
+import org.springframework.util.Assert;
+
+/**
+ * @author Vivek Babu
+ */
+public class OAuth2TokenRevocationAuthenticationProvider implements AuthenticationProvider {
+
+	public static final String TOKEN_TYPE_HINT = "token_type_hint";
+	public static final String UNSUPPORTED_TOKEN_TYPE = "unsupported_token_type";
+
+	private RegisteredClientRepository registeredClientRepository;
+	private OAuth2AuthorizationService authorizationService;
+
+	/**
+	 * Constructs an {@code OAuth2TokenRevocationAuthenticationProvider} using the provided parameters.
+	 *
+	 * @param registeredClientRepository the repository of registered clients
+	 * @param authorizationService the authorization service
+	 */
+	public OAuth2TokenRevocationAuthenticationProvider(RegisteredClientRepository registeredClientRepository,
+			OAuth2AuthorizationService authorizationService) {
+		Assert.notNull(registeredClientRepository, "registeredClientRepository cannot be null");
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.registeredClientRepository = registeredClientRepository;
+		this.authorizationService = authorizationService;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		OAuth2TokenRevocationAuthenticationToken authenticationToken = (OAuth2TokenRevocationAuthenticationToken)
+				authentication;
+
+		String clientId = authenticationToken.getPrincipal().toString();
+		RegisteredClient registeredClient = this.registeredClientRepository.findByClientId(clientId);
+
+		if (registeredClient == null) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT));
+		}
+
+		final String tokenTypeHint = authenticationToken.getTokenTypeHint();
+		final String token = authenticationToken.getToken();
+		final OAuth2Authorization authorization = findAuthorizationUsingHint(tokenTypeHint, token);
+
+		OAuth2TokenRevocationAuthenticationToken successfulAuthentication =
+				new OAuth2TokenRevocationAuthenticationToken(token, registeredClient, tokenTypeHint);
+
+		if (authorization == null) {
+			return successfulAuthentication;
+		}
+
+		if (!registeredClient.getClientId().equals(authorization.getRegisteredClientId())) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_GRANT));
+		}
+
+		if (token.equals(authorization.getAttribute(OAuth2AuthorizationAttributeNames.CODE))) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(UNSUPPORTED_TOKEN_TYPE));
+		}
+
+		this.authorizationService.invalidate(authorization);
+		return successfulAuthentication;
+	}
+
+	private OAuth2Authorization findAuthorizationUsingHint(String tokenTypeHint, String token) {
+		OAuth2Authorization authorization = null;
+		TokenType tokenType = new TokenType(tokenTypeHint);
+
+		if (TokenType.REFRESH_TOKEN.equals(tokenType) || TokenType.ACCESS_TOKEN.equals(tokenType)) {
```

Comment:
Apologies as I did mention to allow the capability to revoke a refresh token but at the moment we don't have refresh token grant feature. Therefore, please remove all references to refresh token. We will add this in after we have refresh token feature implemented.

For now, we can assume the revocation request is for access tokens only. Any other token type hint should return 400 for now.

Reply:
FiJoe the spec says that hint can be any value. if it is of value access_token or refresh_token the authorization server CAN use it to speed up its lookup. If any other value of token_type_hint is give it should not lead to any error

"An invalid token type hint value is ignored by the authorization
   server and does not influence the revocation response."